File: src/sequence_demo_cpp/src/sequence_caller.cpp
#include <chrono>
#include <thread>
#include <vector>

#include "lhandpro_interfaces/srv/home_motors.hpp"
#include "lhandpro_interfaces/srv/move_motors.hpp"
#include "lhandpro_interfaces/srv/set_enable.hpp"
#include "lhandpro_interfaces/srv/set_position.hpp"
#include "lhandpro_interfaces/srv/set_position_velocity.hpp"
#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;

class SequenceCaller : public rclcpp::Node {
 public:
  using SetEnable = lhandpro_interfaces::srv::SetEnable;
  using SetPosition = lhandpro_interfaces::srv::SetPosition;
  using SetPositionVelocity = lhandpro_interfaces::srv::SetPositionVelocity;
  using MoveMotors = lhandpro_interfaces::srv::MoveMotors;
  using HomeMotors = lhandpro_interfaces::srv::HomeMotors;

  SequenceCaller() : Node("sequence_caller") {
    RCLCPP_INFO(this->get_logger(), "Sequence Caller 启动...");

    // 创建客户端
    set_enable_client_ = this->create_client<SetEnable>("set_enable");
    set_position_client_ = this->create_client<SetPosition>("set_position");
    set_position_velocity_client_ =
        this->create_client<SetPositionVelocity>("set_position_velocity");
    move_motors_client_ = this->create_client<MoveMotors>("move_motors");
    home_motors_client_ = this->create_client<HomeMotors>("home_motors");

    // 等待服务上线
    while (!set_enable_client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "等待 set_enable 服务...");
    }

    while (!set_position_client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "等待 set_position 服务...");
    }

    while (!set_position_velocity_client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "等待 set_position_velocity 服务...");
    }

    while (!move_motors_client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "等待 move_motors 服务...");
    }

    while (!home_motors_client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "等待 home_motors 服务...");
    }

    // 使用线程启动轨迹执行
    std::thread([this]() {
      rclcpp::NodeOptions options;
      auto context = std::make_shared<rclcpp::Context>();
      context->init(0, nullptr);

      // 创建一个独立节点用于执行服务调用
      auto sub_node = rclcpp::Node::make_shared("sequence_sub_node");

      // 执行轨迹
      this->execute_trajectory();
    }).detach();  // 自动释放线程资源（注意内存安全）
  }

 public:
  void execute_trajectory() {
    // 设置默认参数
    std::vector<int> joint_ids = {1, 2, 3, 4, 5, 6};
    int default_velocity_ = 20000;

    // 内嵌动作数组
    const std::vector<std::vector<int>> positions = {
        {0, 0, 0, 0, 0, 0},     {10000, 0, 0, 0, 0, 0}, {0, 10000, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},     {0, 0, 10000, 0, 0, 0}, {0, 0, 0, 10000, 0, 0},
        {0, 0, 0, 0, 10000, 0}, {0, 0, 0, 0, 0, 10000}, {0, 0, 0, 0, 0, 0},
    };

    send_set_enable(0, 1);
    RCLCPP_INFO(this->get_logger(), "正在使能");
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

    send_home_motors(0);
    RCLCPP_INFO(this->get_logger(), "正在回零");
    std::this_thread::sleep_for(std::chrono::milliseconds(5000));

    for (size_t i = 0; i < positions.size(); ++i) {
      RCLCPP_INFO(this->get_logger(), "正在执行第 %lu 步动作...", i + 1);

      for (size_t j = 0; j < joint_ids.size(); ++j) {
        int joint_id = joint_ids[j];
        int position = positions[i][j];

        send_set_position(joint_id, position);
        send_set_position_velocity(joint_id, default_velocity_);
      }

      // 触发所有关节运动
      send_move_motors(0);
      RCLCPP_INFO(this->get_logger(), "第 %lu 步动作", i + 1);
      std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    }

    RCLCPP_INFO(this->get_logger(), "轨迹执行完成");
  }

  void send_set_position(int joint_id, int position) {
    auto req = std::make_shared<SetPosition::Request>();
    req->joint_id = joint_id;
    req->position = position;

    auto future = set_position_client_->async_send_request(req);
    future.wait();
    if (future.get()->result == 0) {
      RCLCPP_DEBUG(this->get_logger(), "设置位置成功: %d %d", joint_id,
                   position);
    } else {
      RCLCPP_ERROR(this->get_logger(), "设置位置失败: %d", joint_id);
    }
  }

  void send_set_position_velocity(int joint_id, int velocity) {
    auto req = std::make_shared<SetPositionVelocity::Request>();
    req->joint_id = joint_id;
    req->velocity = velocity;

    auto future = set_position_velocity_client_->async_send_request(req);
    future.wait();
    if (future.get()->result == 0) {
      RCLCPP_DEBUG(this->get_logger(), "设置速度成功: %d %d", joint_id,
                   velocity);
    } else {
      RCLCPP_ERROR(this->get_logger(), "设置速度失败: %d", joint_id);
    }
  }

  void send_set_enable(int joint_id, int enable) {
    auto req = std::make_shared<SetEnable::Request>();
    req->joint_id = joint_id;
    req->enable = enable;

    auto future = set_enable_client_->async_send_request(req);
    future.wait();
    if (future.get()->result == 0) {
      RCLCPP_DEBUG(this->get_logger(), "设置使能成功: %d %d", joint_id, enable);
    } else {
      RCLCPP_ERROR(this->get_logger(), "设置使能失败: %d", joint_id);
    }
  }

  void send_move_motors(int joint_id) {
    auto req = std::make_shared<MoveMotors::Request>();
    req->joint_id = joint_id;

    auto future = move_motors_client_->async_send_request(req);
    future.wait();
    if (future.get()->result != 0) {
      RCLCPP_ERROR(this->get_logger(), "驱动关节失败: %d", joint_id);
    }
  }

  void send_home_motors(int joint_id) {
    auto req = std::make_shared<HomeMotors::Request>();
    req->joint_id = joint_id;

    auto future = home_motors_client_->async_send_request(req);
    future.wait();
    if (future.get()->result != 0) {
      RCLCPP_ERROR(this->get_logger(), "回零关节失败: %d", joint_id);
    }
  }

  rclcpp::Client<SetEnable>::SharedPtr set_enable_client_;
  rclcpp::Client<SetPosition>::SharedPtr set_position_client_;
  rclcpp::Client<SetPositionVelocity>::SharedPtr set_position_velocity_client_;
  rclcpp::Client<MoveMotors>::SharedPtr move_motors_client_;
  rclcpp::Client<HomeMotors>::SharedPtr home_motors_client_;

  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SequenceCaller>());
  rclcpp::shutdown();
  return 0;
}

File: src/sequence_demo_py/sequence_demo_py/sequence_caller.py
import rclpy
from rclpy.node import Node
import os
import time

# 导入服务接口
from lhandpro_interfaces.srv import SetEnable, SetPosition, SetPositionVelocity, MoveMotors, HomeMotors


class SequenceCallerClient(Node):
    def __init__(self):
        super().__init__('sequence_caller')

        # 创建服务客户端
        self.set_enable_cli = self.create_client(SetEnable, 'set_enable')
        self.set_position_cli = self.create_client(SetPosition, 'set_position')
        self.set_position_velocity_cli = self.create_client(SetPositionVelocity, 'set_position_velocity')
        self.move_motors_cli = self.create_client(MoveMotors, 'move_motors')
        self.home_motors_cli = self.create_client(HomeMotors, 'home_motors')

        # 等待服务上线
        while not self.set_enable_cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('等待 set_enable 服务上线...')

        while not self.set_position_cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('等待 set_position 服务上线...')

        while not self.set_position_velocity_cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('等待 set_position_velocity 服务上线...')

        while not self.move_motors_cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('等待 move_motors 服务上线...')

        while not self.home_motors_cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('等待 home_motors 服务上线...')

        # 执行动作序列
        self.execute_trajectory()

    def set_enable(self, joint_id, enable):
        req = SetEnable.Request()
        req.joint_id = joint_id
        req.enable = enable
        future = self.set_enable_cli.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        if future.result() is not None and future.result().result == 0:
            self.get_logger().debug(f'关节 {joint_id} 设置使能成功')
        else:
            self.get_logger().error(f'设置使能失败 (joint_id={joint_id})')

    def set_position(self, joint_id, position):
        req = SetPosition.Request()
        req.joint_id = joint_id
        req.position = position
        future = self.set_position_cli.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        if future.result() is not None and future.result().result == 0:
            self.get_logger().debug(f'关节 {joint_id} 设置位置 {position} 成功')
        else:
            self.get_logger().error(f'设置位置失败 (joint_id={joint_id}, position={position})')

    def set_position_velocity(self, joint_id, velocity):
        req = SetPositionVelocity.Request()
        req.joint_id = joint_id
        req.velocity = velocity
        future = self.set_position_velocity_cli.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        if future.result() is not None and future.result().result == 0:
            self.get_logger().debug(f'关节 {joint_id} 设置速度 {velocity} 成功')
        else:
            self.get_logger().error(f'设置速度失败 (joint_id={joint_id}, velocity={velocity})')

    def move_motors(self, joint_id):
        req = MoveMotors.Request()
        req.joint_id = joint_id
        future = self.move_motors_cli.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        if future.result() is not None and future.result().result == 0:
            self.get_logger().debug(f'关节 {joint_id} 下发运动成功')
        else:
            self.get_logger().error(f'下发运动失败 (joint_id={joint_id})')
    
    def home_motors(self, joint_id):
        req = HomeMotors.Request()
        req.joint_id = joint_id
        future = self.home_motors_cli.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        if future.result() is not None and future.result().result == 0:
            self.get_logger().debug(f'关节 {joint_id} 下发回零成功')
        else:
            self.get_logger().error(f'下发回零失败 (joint_id={joint_id})')

    def execute_trajectory(self):
        # 设置默认参数
        joint_ids = [1, 2, 3, 4, 5, 6]
        default_velocity = 20000

        # 内嵌动作数组
        positions = [
            [0, 0, 0, 0, 0, 0],
            [10000, 0, 0, 0, 0, 0],
            [0, 10000, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0],
            [0, 0, 10000, 0, 0, 0],
            [0, 0, 0, 10000, 0, 0],
            [0, 0, 0, 0, 10000, 0],
            [0, 0, 0, 0, 0, 10000],
            [0, 0, 0, 0, 0, 0],
        ]

        self.set_enable(0, 1)
        self.get_logger().info("正在使能")
        time.sleep(1.0)  # 等待 1 秒

        self.home_motors(0)
        self.get_logger().info("正在回零")
        time.sleep(5)  # 等待 5 秒


        for i, position_step in enumerate(positions, start=1):
            self.get_logger().info(f"正在执行第 {i} 步动作...")

            # 遍历每个关节
            for j, joint_id in enumerate(joint_ids):
                position = position_step[j]
                self.set_position(joint_id, position)
                self.set_position_velocity(joint_id, default_velocity)

            # 触发所有关节运动
            self.move_motors(0)
            self.get_logger().info(f"第 {i} 步动作已触发")
            time.sleep(1.0)  # 等待 1 秒

        self.get_logger().info("轨迹执行完成")


def main(args=None):
    rclpy.init()

    client = SequenceCallerClient()

    client.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()